text/0003-font-types.md

```diff
@@ -0,0 +1,408 @@
+network without any additional storage. Doing this for each individual table,
+however, adds additional complexity.
+
+## Plain old data structures
+
+Another very compelling alternative is to just use plain old Rust data
+structures.
+
+In this world, our macros generate *two* types for each 'shape' that we define.
+One of these is the type that we've already seen, (methods for accessing data
+directly from underlying bytes) but the other one is just a plain old Rust
+struct.
+
+We can bikeshed names later; for the purpose of this illustration let's adopt
+the pattern of naming these new types objects `${TYPE}Owned` where `${TYPE}` is
+the name of the previously defined object. So continuing our `avar` example, we
+get:
+
+```rust
+// identical to above
+struct Avar<'a>(..);
+struct SegmentMaps<'a>(..);
+
+struct AvarOwned {
+    pub version_major: u16,
+    pub version_minor: u16,
+    pub segment_maps: Vec<SegmentMapsOwned>,
+}
+
+struct SegmentMapsOwned(pub Vec<AxisValueMap>);
+```
+
+All of these 'Owned' types would be plain Rust structs (and likely enums)
+generated by the macros, and they would also get generated code for writing them
+back out to bytes.
+
+This has the advantage of being a trivially easy API without any visible
+fanciness. You would describe your tables using simple rust types, using `std`
+APIs to mutate values, modify collections, or otherwise manipulate them. When
+you were happy with your representation, you would encode once, and then have
+the valid binary representation of your table (and you could still reuse the
+data structure, for instance modifying it further and writing out a different
+version).
+
+The downside of this is that there are now two types for each non-scalar
+structure in the spec; and additionally you are holding on to two
+representations of the data. In practice I don't mind this too much. One way to
+think about this is that these data structures are basically just another form
+of builder, that happen to be reusable and that maintain some intermediate
+state.
+
+### relationships between these two types
+
+Initially, it may be useful to think of these two sets of types as basically
+independent of one another, but ultimately we will want some mechanisms for
+moving between them.
+
+I think for the time being we can keep this simple, though: we can have simple
+equivalents of the `From` and `Into` traits. If we have an existing table that
+we want to modify, we can convert it into the decoded form, modify it
+arbitrarily, and then re-encode it.
+
+## Fancier options
+
+My personal current preference is for the second option of creating two versions
+of each table; my hunch is that this will end up being a much simpler API in the
+general case. In most cases, the user will be using one of these two sets of
+types, and in the advanced cases (e.g. server-side subsetting, instancing
+variable fonts) it still shouldn't be that complex.
```

Comment:
Can we elaborate on how server-side subsetting would work, e.g. how would wanting to efficiently produce many subsets from the same readonly instance of a large font, potentially concurrently, play out?

Reply:
To do this justice I would need to have a better understanding of how subsetting is currently implemented. I think this would be a useful exercise, though, and would help me better understand some of the possible constraints. If there's any public documentation someone could point me to, that would be helpful.

Reply:
The [Faster Horse](https://docs.google.com/document/d/1vDbMhZFkj52F1YbG7X7KeHCc1VUB7hRHgsx4rYwBQls/edit#heading=h.g7d19owvbzeb) design doc is a good overview. Scaling primarily based on the number of glyphs in the **output** is particularly key, anything that scales significantly with glyphs in input will fare poorly for CJK and that is a critical use case for us.

https://github.com/harfbuzz/harfbuzz/blob/main/docs/repacker.md covers the ugliest part. 

@garretrieger any other key refs?

Reply:
In addition to the implementation, it would also be useful to have a sense of how this is deployed. For instance it might make sense to have some kind of long-lived 'SubsettingContext' that handles multiple requests, and reuses allocations, or it might make sense to use an arena allocator. If we wanted to support cases like this it might be worth thinking about that now?